src/worker.cc

```diff
@@ -983,13 +984,20 @@ void from_json(rapidjson::Document& doc, Options::Action action, Api& api) {
       rapidjson::get_child_optional(doc, doc.HasMember("avoid_polygons") ? "/avoid_polygons"
                                                                          : "/exclude_polygons");
   if (rings_req) {
-    auto* rings_pbf = options.mutable_exclude_polygons();
-    try {
-      for (const auto& req_poly : rings_req->GetArray()) {
-        auto* ring = rings_pbf->Add();
-        parse_ring(ring, req_poly);
```

Comment:
what was wrong with this, it should never segfault. we purposely force rapidjson to throw exceptions instead of abort. to my knowledge the only thing in here we do is use rapidjson to parse stuff with like GetArray and GetDouble or whatever, those should all throw when the types are not correct and it should fall into the exception below that says dont send us junk input. if thats not happening either its not rapidjson parsing that is failing or we screwed up the whole making it throw instead of abort stuff in rapidjson utils at some point. if i revert this and run the new tests does it segfault?

Reply:
> if i revert this and run the new tests does it segfault?

Jep, it will. GTest is outputting
```
[build] make[3]: *** [test/gurka/CMakeFiles/run-gurka_avoids.dir/build.make:73: test/gurka/gurka_avoids.log] Error 139
[build] make[3]: *** Deleting file 'test/gurka/gurka_avoids.log'
[build] make[2]: *** [CMakeFiles/Makefile2:11115: test/gurka/CMakeFiles/run-gurka_avoids.dir/all] Error 2
[build] make[1]: *** [CMakeFiles/Makefile2:11122: test/gurka/CMakeFiles/run-gurka_avoids.dir/rule] Error 2
[build] make: *** [Makefile:4609: run-gurka_avoids] Error 2
```

It's the polygon_search that segfaults. In the loki worker we send the polygons to `polygon_search` if they're not empty:
https://github.com/valhalla/valhalla/blob/d182120b4312b2534d15927d734e4a4a7a688808/src/loki/worker.cc#L87-L89

Then in polygon_search it segfaults when it wants to correct a polygon: the uninitialized `new_ring` is passed into `correct_ring()` which doesn't have any elements and it segfaults on the `bool is_open = (ring.begin()->lat() != ring.rbegin()->lat() || ring.begin()->lng() != ring.rbegin()->lng());` line.

https://github.com/valhalla/valhalla/blob/d182120b4312b2534d15927d734e4a4a7a688808/src/loki/polygon_search.cc#L36-L57

IMO catching the problem already while parsing the input JSON is best, so it never gets here. But thinking about it again, it's probably a good idea to protect the polygon_search as well, as it's public API and could affect downstream programs in the exact same way. I'll quickly add that and we can discuss.

Reply:
I agree the sooner we see bogus input and throw the better but protecting downstream library functions must also be fixed as you point out